Orthogonal-view-in-editor

SYBIOTE opened the pull request:
PR: Enhanced Camera Tool Icons and Anchor-Relative Positioning
Summary
Adds new grid icons for camera projection modes and improves draggable positioning with anchor-relative coordinates.
Changes
New Icons: Grid3x3Sm (orthographic) and GridPerspectiveSm (perspective) with consistent line thickness
Enhanced useDraggable: targetStartX/Y now relative to anchor position (e.g., right anchor positions from right edge)
Updated SceneCameraTool: Uses anchor: 'right' with targetStartX: 20 for better positioning

SYBIOTE committed:
Update dependencies and enhance draggable functionality

- Added polyfills for Node globals and modules in `vite.config.ts`.
- Updated `package.json` to include new dependencies: `@esbuild-plugins/node-globals-polyfill`, `@esbuild-plugins/node-modules-polyfill`, `buffer`, and `process`.
- Enhanced `useDraggable` hook to support anchor positioning, allowing elements to be positioned relative to specified anchor points.
- Improved camera handling in various components to ensure compatibility with both perspective and orthographic projections.
- Fixed camera frustum and body geometry calculations to correctly handle Z-axis flipping.
- Updated viewport to include `SceneCameraTool` for enhanced scene management.

SYBIOTE committed:
Refactor SceneCameraTool to use new grid icons

- Replaced Grid01Sm and GridDotsHorizontalCenterSm with Grid3x3Sm and GridPerspectiveSm in SceneCameraTool for improved visual representation.
- Added new grid icons to the icons index for better accessibility in the UI.

AidanCaruso reviewed (changes requested):
<img width="827" height="324" alt="Image" src="https://github.com/user-attachments/assets/9e4a6f69-44ba-4529-98c9-f82779fd03a5" />

Looks like the camera perspective button is not right aligned. Nice work fixing the polyfill issue and unblocking the studio route.

SYBIOTE commented:
It gets pushed inward when the right edge comes in, but it retains the position when pulling out, 
I encountered and seemingly fixed it, ill check again

SYBIOTE committed:
Enhance useDraggable hook with initial anchor distance support

- Added initialAnchorDistanceX and initialAnchorDistanceY parameters to the useDraggable hook for improved positioning.
- Refactored anchor handling logic to maintain relative positions based on initial anchor distances.
- Updated the constrainToParent and maintainAnchorPosition functions to utilize the new anchor distance features.
- Improved overall drag-and-drop functionality by ensuring elements remain anchored correctly during interactions.

SYBIOTE committed:
Refactor camera handling to support ArrayCamera type

- Updated imports to include ArrayCamera in various components and systems.
- Modified camera-related logic to cast CameraComponent to ArrayCamera where necessary.
- Enhanced rendering functions to ensure compatibility with ArrayCamera, improving overall camera functionality.
- Adjusted tests to reflect changes in camera handling, ensuring accurate assertions for ArrayCamera properties.

SYBIOTE committed:
Refactor camera handling in various components to ensure type safety and compatibility

- Removed unnecessary type casting for camera components, explicitly using PerspectiveCamera and ArrayCamera where applicable.
- Updated screenshot functions to correctly handle camera types, improving reliability in thumbnail generation.
- Cleaned up variable names for clarity in the ActiveHelperSystem.
- Enhanced the CameraNodeEditor to ensure proper handling of camera properties.

SYBIOTE committed:
Refactor camera systems to improve projection handling and type safety

- Removed debug logging from SceneCameraTool for cleaner code.
- Updated CameraOrbitSystem to eliminate unnecessary orthographic camera handling logic.
- Introduced separate functions for updating orthographic and perspective cameras in CameraSystem, enhancing clarity and maintainability.
- Ensured proper aspect ratio handling for both camera types during rendering.

SYBIOTE committed:
Enhance camera type handling and projection logic across components

- Introduced type guards for Orthographic and Perspective cameras to improve type safety.
- Updated camera handling in various systems to utilize new type guards, ensuring correct camera type usage.
- Refactored rendering functions to accommodate both camera types, enhancing clarity and maintainability.
- Improved aspect ratio and near/far plane handling for better rendering performance.

SYBIOTE committed:
Merge branch 'dev' into orthogonal-view-in-editor

SYBIOTE committed:
fix camera properties panel for orthogonal camera

SYBIOTE committed:
fix linter errors

SYBIOTE committed:
Merge branch 'dev' into orthogonal-view-in-editor

SYBIOTE committed:
fix camera properties for ecs optimizations

SYBIOTE committed:
fix camera component

SYBIOTE committed:
ecs changes for Camera System

SYBIOTE committed:
split into two camera components

SYBIOTE committed:
fix file name

SYBIOTE committed:
Merge branch 'dev' into orthogonal-view-in-editor

SYBIOTE committed:
initialize camera synchrounously

SYBIOTE committed:
ensure perspective camera assignment

SYBIOTE committed:
remove ununeeded functions

HexaField committed:
Merge branch 'dev' into orthogonal-view-in-editor

HexaField commented on packages/spatial/src/camera/components/CameraComponent.ts:
```diff
@@ -1,35 +1,31 @@
 import { useImmediateEffect } from '@ir-engine/hyperflux'
 import { ObjectComponent } from '../../renderer/components/ObjectComponent'
 
+export type CameraComponentType = ArrayCamera | OrthographicCamera
```
The camera component should still be a camera array. It'll add some complexity but it keeps things isomorphic.

SYBIOTE replied:
noted
